Support domain wildcards/regex via if-frame-url triggers

hyeonjongyang opened the pull request:
Safari 26 adds `if-frame-url` / `unless-frame-url` triggers. This PR uses a hybrid approach to support `domain.*` and domain-regex values while keeping the fast path (`if-domain`/`unless-domain`) for plain domains to avoid unnecessary compilation overhead (addresses #100).

## Changes:
- For Safari 26+:
  - Keep using `if-domain` / `unless-domain` for plain domains.
  - Use `if-frame-url` / `unless-frame-url` only for:
     - TLD wildcards like `example.*`
     - Domain regexes like `/test\\.(com|net)/`
  - If a rule contains both plain + special domains, split it into multiple entries (one per trigger type) with the same action/other trigger fields.
- Update domain list parsing so `|` inside `/.../` domain regex doesn’t split the domain list.
- Preserve older-Safari third-party workaround behavior.
- Add unit tests and a changelog entry.

hyeonjongyang committed:
Support domain wildcards/regex via if-frame-url triggers

hyeonjongyang committed:
Fix empty $domain regex validation

hyeonjongyang committed:
Fix regex-domain closing slash with escaped backslashes

ameshkov commented on Sources/ContentBlockerConverter/Compiler/BlockerEntryFactory.swift:
```diff
@@ -443,28 +437,154 @@ class BlockerEntryFactory {
+            if isTldWildcardDomain(domain) {
+                // Convert `example.*` to a regex that matches `example.<any suffix>`.
+                let prefix = String(domain.dropLast(2))
+                let escaped = NSRegularExpression.escapedPattern(for: prefix)
```
Looks like an overkill for simply replacing `.` with `\.`

ameshkov commented on Sources/ContentBlockerConverter/Compiler/BlockerEntryFactory.swift:
```diff
@@ -443,28 +437,154 @@ class BlockerEntryFactory {
+
+        if !plain.isEmpty {
+            var trigger = baseTrigger
+            let wildcardDomains = plain.map { "*" + $0 }
```
```suggestion
            // Prepend * to include subdomains.
            let wildcardDomains = plain.map { "*" + $0 }
```

ameshkov reviewed (changes requested):
After merging the previous pull request there are some conflicts to resolve.

In addition, there are some changes that needs to be done for any PR like that. For guidance please see the merge commit with changes that was made to your previous pull request: https://github.com/AdguardTeam/SafariConverterLib/commit/def0b10bf47c3f1a6d7bda2d69cf24fa4849f77a

Here's what needs to be done:

1. README should be updated with the new information on compatibility
2. Please run `make lint` to make sure that the code is formatted correctly.
3. Please update the relevant bench tests with your baseline before and after the change.
4. `FilterEngine` and `FilterRule` at this point don't need to support TLD/regexes, but we'll need a separate issue for this.

hyeonjongyang committed:
Validate Safari 26 frame-url patterns

hyeonjongyang committed:
Update tests for Safari 26 frame-url domains

hyeonjongyang committed:
Update README compatibility and benchmarks

hyeonjongyang committed:
Update CosmeticRule formatting

hyeonjongyang committed:
Fix NetworkRule modifier parsing

hyeonjongyang committed:
Fix CosmeticRuleTests for Safari 26 regex domains

hyeonjongyang committed:
Update performance baseline comments

ameshkov commented on Sources/ContentBlockerConverter/Compiler/BlockerEntryFactory.swift:
```diff
@@ -469,27 +469,244 @@ class BlockerEntryFactory {
+        return (plain, special)
+    }
+
+    private func isRegexDomain(_ domain: String) -> Bool {
```
I think this function is not needed anymore

ameshkov commented on Sources/ContentBlockerConverter/Compiler/BlockerEntryFactory.swift:
```diff
@@ -469,27 +469,244 @@ class BlockerEntryFactory {
+    }
+
+    private func unescapeDomainRegex(_ pattern: String) -> String {
+        // In `$domain=/regexp/` the following characters must be escaped in the filter syntax:
```
Please move the explanation to the function comment for more clarity

ameshkov commented on Sources/ContentBlockerConverter/Compiler/BlockerEntryFactory.swift:
```diff
@@ -469,27 +469,244 @@ class BlockerEntryFactory {
+        return domain.hasSuffix(".*")
+    }
+
+    private func unescapeDomainRegex(_ pattern: String) -> String {
```
We will need to use the same function when adding support for domain regexes to `FilterEngine` so I suggest moving this function to `SimpleRegex` and adding unit tests for it.

ameshkov commented on Sources/ContentBlockerConverter/Compiler/BlockerEntryFactory.swift:
```diff
@@ -469,27 +469,244 @@ class BlockerEntryFactory {
+                let pattern = #"^[^:]+://+([^:/]+\.)?\#(escaped)\.[^/:]+([/:?#].*)?$"#
+                try validateSafariRegex(pattern: pattern, context: domain)
+                result.append(pattern)
+                continue
```
I prefer to avoid using `continue` when possible as it makes it harder to understand the function flow.

Please refactor into smaller functions so that the flow looked like this:

```
if isTldWildcardDomain(domain) {
    let pattern = try createTldWildcardPattern(domain)
    result.append(pattern)
} else if ... {
    let pattern = try createRegexDomainPattern   
} else {
    throw ....
}
```

ameshkov commented on Tests/ContentBlockerConverterTests/ContentBlockerConverterPerformanceTests.swift:
```diff
@@ -64,6 +63,12 @@ extension ContentBlockerConverterTests {
     /// - Swift: 6.2
     /// - Average execution time: ~0.787 sec
     ///
+    /// Baseline results (Dec 25, 2025):
```
Could you please check performance before the change and tell me what's the difference?

ameshkov reviewed (changes requested):
Thank you!

Sorry that the review takes so long, this is due to New Year vacation.

Overall, looks good to me, there are just a couple of things to address.

hyeonjongyang committed:
Unescape $domain regex patterns in SimpleRegex

Move $domain=/regexp/ unescaping into SimpleRegex.unescapeDomainRegex and add unit tests.

hyeonjongyang committed:
Refactor Safari 26 domain trigger building

Split frame-url domain pattern building into helpers, remove continue-based control flow, and document legacy Safari behavior.

hyeonjongyang committed:
Update Safari 26 test constants

hyeonjongyang committed:
Remove PR performance check comments

hyeonjongyang committed:
Fix lint error

hyeonjongyang commented:
Forgot to run the lint 😅
